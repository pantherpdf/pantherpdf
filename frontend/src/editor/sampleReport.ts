import { TReport } from '../../../backend/shared/types'

export const sampleReport: TReport = {
	_id: 'abc-def-123',
	name: 'Report 1',
	email: 'admin@test.com',
	time: '2021-07-01T13:00:00Z',
	children: [],
	target: 'pdf',
	properties: {
	},
}
