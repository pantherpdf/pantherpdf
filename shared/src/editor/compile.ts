/**
 * compile.ts
 * Compile sourceData and report - evaluate all formulas
 */

import { TReport, TData } from '../types'
import type { TReportCompiled, TDataCompiled } from './types'
import FormulaEvaluate from '../formula/formula'
import getWidget from '../widgets/allWidgets'

type TOvrr = [string, any]

export class Helper {
	overrides: TOvrr[]

	constructor() {
		this.overrides = []
	}

	push(name: string, valueOrCb: any) {
		this.overrides.push([name, valueOrCb])
	}

	pop() {
		this.overrides.pop()
	}

	async getVar(name: string): Promise<any> {
		for (let i=this.overrides.length-1; i>=0; --i) {
			if (this.overrides[i][0] == name) {
				let val = this.overrides[i][1]
				if (typeof val == 'function') {
					val = await val()
				}
				return val
			}
		}
		return undefined
	}

	async evalFormula(txt: string): Promise<any> {
		return FormulaEvaluate(txt, {getVar: this.getVar.bind(this)})
	}

	async compileChildren(arr1: TData[]): Promise<TDataCompiled[]> {
		// dont use promise.all() because order of execution matters and some async operation could change it
		const arr2 = []
		for (const ch of arr1) {
			const dt = await getWidget(ch.type).compile(ch, this)
			arr2.push(dt)
		}
		return arr2
	}
}


export default async function compile(dt: TReport, obj: any): Promise<TReportCompiled> {
	const helper = new Helper()
	helper.push('obj', obj)
	const dt2: TReportCompiled = {...dt}
	dt2.children = await helper.compileChildren(dt.children)
	helper.pop()	
	return dt2
}
