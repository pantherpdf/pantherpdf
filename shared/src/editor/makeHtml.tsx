/**
 * makeHtml.ts
 * Compiled report -> html
 */


import React from 'react'
import type { TReportCompiled, ItemRendeFinalProps, TDataCompiled, ItemRendeFinalHelper } from './types'
import getWidget from '../widgets/allWidgets'


export default function makeHtml(report: TReportCompiled) {
	const helper: ItemRendeFinalHelper = {
		renderItem: (item, helper) => {
			const w = getWidget(item.type)
			return <w.RenderFinal {...helper} item={item} />
		},
		renderChildren: (chs, helper) => {
			return <div>{chs.map((item, idx) => {
				const w = getWidget(item.type)
				return <w.RenderFinal {...helper} item={item} key={idx} />
			})}</div>  // todo key
		},
	}

	const arr = helper.renderChildren(report.children, helper)
	return arr
}
