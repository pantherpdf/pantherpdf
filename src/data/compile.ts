/**
 * @file Compile sourceData and report - evaluate all formulas
 * @project PantherPDF Report Editor
 * @copyright Ignac Banic 2021-2024
 * @license MIT
 */

import type {
  Report,
  ApiEndpoints,
  ReportCompiled,
  ReportProperties,
  ReportPropertiesCompiled,
  FormulaObject,
} from '../types';
import formulaEvaluate from '../formula/formula';
import { getWidget } from '../widgets/allWidgets';
import type { CompileHelper, Widget, WidgetItem } from '../widgets/types';
import { PropertyFontExtractStyle } from '../widgets/PropertyFont';

type TOvrr = [string, unknown];

export class FormulaHelper {
  overrides: TOvrr[];

  constructor() {
    this.overrides = [];
  }

  push(name: string, valueOrCb: unknown) {
    this.overrides.push([name, valueOrCb]);
  }

  pop() {
    this.overrides.pop();
  }

  async getVar(name: string): Promise<unknown> {
    for (let i = this.overrides.length - 1; i >= 0; --i) {
      if (this.overrides[i][0] === name) {
        let val = this.overrides[i][1];
        if (typeof val == 'function') {
          val = await val(name);
        }
        return val;
      }
    }
    return undefined;
  }
}

export default async function compile(
  report: Report,
  data: unknown,
  widgets: Widget[],
  api: ApiEndpoints,
  externalHelpers: { [key: string]: unknown } = {},
): Promise<ReportCompiled> {
  // make a copy, to support changing
  report = JSON.parse(JSON.stringify(report));

  const formulaHelper = new FormulaHelper();
  formulaHelper.push('data', data);
  formulaHelper.push('report', report);
  const evalFormulaWrapper = async (val: FormulaObject): Promise<unknown> => {
    return formulaEvaluate(val.formula, formulaHelper);
  };

  // custom variables
  const vars: { [key: string]: unknown } = {};
  function getVarValue(varName: string): unknown {
    return vars[varName];
  }
  for (const v of report.variables) {
    vars[v.name] = evalFormulaWrapper(v.value);
    formulaHelper.push(v.name, getVarValue);
  }

  const propsCompiled = await compilePrprt(
    report.properties,
    evalFormulaWrapper,
  );

  const helper: CompileHelper = {
    wid: [],
    report: report,
    propertiesCompiled: propsCompiled,
    formulaHelper,
    api,
    variables: vars,
    externalHelpers,

    evalFormula: evalFormulaWrapper,

    compileChildren: async (children: WidgetItem[], helper: CompileHelper) => {
      // dont use promise.all() because order of execution matters and some async operation could change it
      const arr2 = [];
      for (let i = 0; i < children.length; ++i) {
        const ch = children[i];
        const helper2 = { ...helper, wid: [...helper.wid, i] };
        const dt = await getWidget(widgets, ch.type).compile(ch, helper2);
        arr2.push(dt);
      }
      return arr2;
    },
  };

  const widgetsCompiled = await helper.compileChildren(report.widgets, helper);

  formulaHelper.pop();
  formulaHelper.pop();
  report.variables.map(() => formulaHelper.pop());
  if (formulaHelper.overrides.length !== 0) {
    throw new Error('helper has overrides still left inside');
  }

  return {
    widgets: widgetsCompiled,
    properties: propsCompiled,
  };
}

async function compilePrprt(
  prprt: ReportProperties,
  evalFormulaWrapper: (formula: FormulaObject) => Promise<unknown>,
): Promise<ReportPropertiesCompiled> {
  const obj: ReportPropertiesCompiled = {
    // make deep copy because widgets can change properties
    ...JSON.parse(JSON.stringify(prprt)),
    fileName: undefined,
    fontsUsed: [],
    globalCss: '',
  };
  if (prprt.fileName !== undefined) {
    const res = await evalFormulaWrapper(prprt.fileName);
    if (typeof res !== 'string') {
      throw new Error(
        `Filename should be a string, but received ${typeof res}`,
      );
    }
    obj.fileName = res;
  }
  if (prprt.font) {
    const style = PropertyFontExtractStyle(prprt.font);
    if (style) {
      obj.fontsUsed.push(style);
    }
  }
  return obj;
}
