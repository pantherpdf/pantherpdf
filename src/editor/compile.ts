/**
 * @file Compile sourceData and report - evaluate all formulas
 * @project PantherPDF Report Editor
 * @copyright Ignac Banic 2021
 * @license MIT
 */

import { Report, Item, ApiEndpoints } from '../types';
import type { ReportCompiled } from '../types';
import FormulaEvaluate from '../formula/formula';
import { getWidget } from '../widgets/allWidgets';
import { CompileHelper, Widget } from './types';
import { PropertyFontExtractStyle } from '../widgets/PropertyFont';

type TOvrr = [string, unknown];

export class FormulaHelper {
  overrides: TOvrr[];

  constructor() {
    this.overrides = [];
  }

  push(name: string, valueOrCb: unknown) {
    this.overrides.push([name, valueOrCb]);
  }

  pop() {
    this.overrides.pop();
  }

  async getVar(name: string): Promise<unknown> {
    for (let i = this.overrides.length - 1; i >= 0; --i) {
      if (this.overrides[i][0] === name) {
        let val = this.overrides[i][1];
        if (typeof val == 'function') {
          val = await val(name);
        }
        return val;
      }
    }
    return undefined;
  }
}

export default async function compile(
  report: Report,
  data: unknown,
  widgets: Widget[],
  api: ApiEndpoints,
  externalHelpers: { [key: string]: any } = {},
): Promise<ReportCompiled> {
  // make a copy, to support changing
  report = JSON.parse(JSON.stringify(report));

  const formulaHelper = new FormulaHelper();
  const getVar = formulaHelper.getVar.bind(formulaHelper);
  formulaHelper.push('data', data);
  formulaHelper.push('report', report);

  // custom variables
  const vars: { [key: string]: unknown } = {};
  function getVarValue(varName: string): unknown {
    return vars[varName];
  }
  for (const v of report.variables) {
    vars[v.name] = await FormulaEvaluate(v.formula, { getVar });
    formulaHelper.push(v.name, getVarValue);
  }

  let dt2: ReportCompiled = {
    ...report,
    children: [],
    fontsUsed: [],
    globalCss: '',
  };
  dt2 = JSON.parse(JSON.stringify(dt2));
  if (dt2.properties.font) {
    const style = PropertyFontExtractStyle(dt2.properties.font);
    if (style) {
      dt2.fontsUsed.push(style);
    }
  }

  if (dt2.properties.fileName) {
    const res = await FormulaEvaluate(dt2.properties.fileName, { getVar });
    if (typeof res !== 'string') {
      throw new Error(
        `Filename should be a string, but received ${typeof res}`,
      );
    }
    dt2.properties.fileName = res;
  }

  const helper: CompileHelper = {
    wid: [],
    report: report,
    reportCompiled: dt2,
    formulaHelper,
    api,
    variables: vars,
    externalHelpers,

    evalFormula: async (txt: string) => {
      return FormulaEvaluate(txt, { getVar });
    },

    compileChildren: async (children: Item[], helper: CompileHelper) => {
      // dont use promise.all() because order of execution matters and some async operation could change it
      const arr2 = [];
      for (let i = 0; i < children.length; ++i) {
        const ch = children[i];
        const helper2 = { ...helper, wid: [...helper.wid, i] };
        const dt = await getWidget(widgets, ch.type).compile(ch, helper2);
        arr2.push(dt);
      }
      return arr2;
    },
  };

  dt2.children = await helper.compileChildren(report.children, helper);

  formulaHelper.pop();
  formulaHelper.pop();
  report.variables.map(() => formulaHelper.pop());
  if (formulaHelper.overrides.length !== 0) {
    throw new Error('helper has overrides still left inside');
  }
  return dt2;
}
