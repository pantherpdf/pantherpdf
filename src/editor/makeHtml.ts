/**
 * makeHtml.ts
 * Compiled report -> html
 */


import type { CSSProperties } from 'react'
import { defaultReportCss, TReportCompiled } from '../types'
import type { ItemRendeFinalHelper } from './types'
import getWidget from '../widgets/allWidgets'
import styleToCssString from 'react-style-object-to-css'
import { PropertyFontGenCss } from '../widgets/PropertyFont'
import { GoogleFontCssUrl } from '../widgets/GoogleFonts'
import { encodeHtml } from '../widgets/HtmlParser'


function escapeHtml(unsafe: string): string {
	return unsafe
		.replace(/&/g, "&amp;")
		.replace(/</g, "&lt;")
		.replace(/>/g, "&gt;")
		.replace(/"/g, "&quot;")
		.replace(/'/g, "&#039;");
}


export function makeHtmlContent(report: TReportCompiled, externalHelpers: {[key: string]: any}={}) {
	const helper: ItemRendeFinalHelper = {
		renderItem: (item, helper) => {
			const w = getWidget(item.type)
			return w.RenderFinal({...helper, item})
		},
		renderChildren: (chs, helper) => {
			if (!Array.isArray(chs)) {
				throw new Error('Bad childs array')
			}
			let txt = ''
			for (const item of chs) {
				const w = getWidget(item.type)
				txt += w.RenderFinal({...helper, item})
			}
			return txt
		},
		escapeHtml,
		styleToStringAttribute: (css: CSSProperties) => escapeHtml(styleToCssString(css).trim()),
		externalHelpers,
	}

	return helper.renderChildren(report.children, helper)
}


export default function makeHtml(report: TReportCompiled, externalHelpers: {[key: string]: any}={}): string {
	// prepare css
	const cssObj: CSSProperties = {...defaultReportCss, ...PropertyFontGenCss(report.properties.font || {})}
	const css = styleToCssString(cssObj)

	// render content
	const htmlContent = makeHtmlContent(report, externalHelpers)

	const fonts = new Set(report.fontsUsed)
	const fontUrls = [...fonts]
	.map(x => GoogleFontCssUrl(x) || '')
	.filter(x => x.length > 0)
	.map(x => `<link rel="stylesheet" href="${encodeHtml(x)}">`)
	.join('\n')

	const html = `<!DOCTYPE html>
<html lang="en-US">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<title>Report</title>
<style>
/* http://meyerweb.com/eric/tools/css/reset/ 
   v2.0 | 20110126
   License: none (public domain)
*/
html, body, div, span, applet, object, iframe,
h1, h2, h3, h4, h5, h6, p, blockquote, pre,
a, abbr, acronym, address, big, cite, code,
del, dfn, em, img, ins, kbd, q, s, samp,
small, strike, strong, sub, sup, tt, var,
b, u, i, center,
dl, dt, dd, ol, ul, li,
fieldset, form, label, legend,
table, caption, tbody, tfoot, thead, tr, th, td,
article, aside, canvas, details, embed, 
figure, figcaption, footer, header, hgroup, 
menu, nav, output, ruby, section, summary,
time, mark, audio, video {
	margin: 0;
	padding: 0;
	border: 0;
	font-size: 100%;
	font: inherit;
	vertical-align: baseline;
}
/* HTML5 display-role reset for older browsers */
article, aside, details, figcaption, figure, 
footer, header, hgroup, menu, nav, section {
	display: block;
}
body {
	line-height: 1.25;
}
ol, ul {
	list-style: none;
}
blockquote, q {
	quotes: none;
}
blockquote:before, blockquote:after,
q:before, q:after {
	content: '';
	content: none;
}
table {
	border-collapse: collapse;
	border-spacing: 0;
}
/* modifications */
b, strong {
	font-weight: bold;
}
cite, em, i {
	font-style: italic;
}
big {
	font-size: larger;
}
small {
	font-size: smaller;
}
mark {
	background-color: yellow;
}
sub {
    vertical-align: sub;
    font-size: smaller;
}
sup {
    vertical-align: super;
    font-size: smaller;
}
</style>
<style>
* {
	-webkit-print-color-adjust: exact;
	print-color-adjust: exact;
}
</style>
<style>
body {
	${css}
}
${report.globalCss}
</style>
${fontUrls}
</head>
<body>
${htmlContent}
</body>
</html>
`
	return html
}
