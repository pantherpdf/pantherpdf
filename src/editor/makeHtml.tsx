/**
 * makeHtml.ts
 * Compiled report -> html
 */


import React from 'react'
import type { TReportCompiled } from '../types'
import type { ItemRendeFinalHelper } from './types'
import getWidget from '../widgets/allWidgets'


export default function makeHtml(report: TReportCompiled, externalHelpers: {[key: string]: any}={}) {
	const helper: ItemRendeFinalHelper = {
		renderItem: (item, helper) => {
			const w = getWidget(item.type)
			return <w.RenderFinal {...helper} item={item} />
		},
		renderChildren: (chs, helper) => {
			if (!Array.isArray(chs)) {
				throw new Error('Bad childs array')
			}
			return <>{chs.map((item, idx) => {
				const w = getWidget(item.type)
				return <w.RenderFinal {...helper} item={item} key={idx} />
			})}</>
		},
		externalHelpers,
	}

	const arr = helper.renderChildren(report.children, helper)
	return arr
}
