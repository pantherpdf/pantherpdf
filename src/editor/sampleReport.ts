/**
 * @project PantherPDF Report Editor
 * @copyright Ignac Banic 2023
 * @license MIT
 */

import type { Report } from '../types';

export const sampleReport: Report = {
  name: 'Report 1',
  widgets: [],
  transforms: [],
  properties: {},
  variables: [],
};
