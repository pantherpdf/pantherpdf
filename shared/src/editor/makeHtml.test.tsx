/**
 * @jest-environment node
 */


import compile from './compile'
import type { ReportForceChildren } from './types'
import type { TextSimpleData, TextSimpleCompiled } from '../widgets/textSimple'
import { sampleReport } from './sampleReport'
import makeHtml from './makeHtml'
import renderer from 'react-test-renderer'

test('text', async () => {
	const report: ReportForceChildren<TextSimpleData> = {
		...sampleReport,
		children: [
			{type:'textSimple', formula:'"Hello World: "+obj.num', children:[]},
		]
	}

	const obj = { num: 123 }
	const compiled = await compile(report, obj)
	const html = makeHtml(compiled)
	const component = renderer.create(html)
	const tree = component.toJSON()
  	expect(tree).toMatchSnapshot();
})
