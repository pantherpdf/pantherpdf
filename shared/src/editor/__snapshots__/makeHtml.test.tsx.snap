// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`text 1`] = `
<div>
  <div>
    Hello World: 123
  </div>
</div>
`;
